client.go

```diff
@@ -222,42 +222,54 @@ func decodeString(body io.Reader, output *string) error {
 	return nil
 }
 
+type fullURLOptions struct {
+	model string
+}
+
+type fullURLOption func(*fullURLOptions)
+
+func withModel(model string) fullURLOption {
+	return func(args *fullURLOptions) {
+		args.model = model
+	}
+}
+
+var azureDeploymentsEndpoints = []string{
+	"/completions",
+	"/embeddings",
+	"/chat/completions",
+	"/audio/transcriptions",
+	"/audio/translations",
+	"/audio/speech",
+	"/images/generations",
+}
+
 // fullURL returns full URL for request.
-// args[0] is model name, if API type is Azure, model name is required to get deployment name.
-func (c *Client) fullURL(suffix string, args ...any) string {
-	// /openai/deployments/{model}/chat/completions?api-version={api_version}
-	if c.config.APIType == APITypeAzure || c.config.APIType == APITypeAzureAD {
-		baseURL := c.config.BaseURL
-		baseURL = strings.TrimRight(baseURL, "/")
-		parseURL, _ := url.Parse(baseURL)
-		query := parseURL.Query()
-		query.Add("api-version", c.config.APIVersion)
-		// if suffix is /models change to {endpoint}/openai/models?api-version=2022-12-01
-		// https://learn.microsoft.com/en-us/rest/api/cognitiveservices/azureopenaistable/models/list?tabs=HTTP
-		if containsSubstr([]string{"/models", "/assistants", "/threads", "/files"}, suffix) {
-			return fmt.Sprintf("%s/%s%s?%s", baseURL, azureAPIPrefix, suffix, query.Encode())
-		}
-		azureDeploymentName := "UNKNOWN"
-		if len(args) > 0 {
-			model, ok := args[0].(string)
-			if ok {
-				azureDeploymentName = c.config.GetAzureDeploymentByModel(model)
-			}
-		}
-		return fmt.Sprintf("%s/%s/%s/%s%s?%s",
-			baseURL, azureAPIPrefix, azureDeploymentsPrefix,
-			azureDeploymentName, suffix, query.Encode(),
-		)
+func (c *Client) fullURL(suffix string, setters ...fullURLOption) string {
+	baseURL := strings.TrimRight(c.config.BaseURL, "/")
+	args := fullURLOptions{}
+	for _, setter := range setters {
+		setter(&args)
 	}
 
-	// https://developers.cloudflare.com/ai-gateway/providers/azureopenai/
-	if c.config.APIType == APITypeCloudflareAzure {
-		baseURL := c.config.BaseURL
-		baseURL = strings.TrimRight(baseURL, "/")
-		return fmt.Sprintf("%s%s?api-version=%s", baseURL, suffix, c.config.APIVersion)
+	if c.config.APIVersion != "" {
+		parseSuffix, _ := url.Parse(suffix)
+		query := parseSuffix.Query()
+		query.Add("api-version", c.config.APIVersion)
+		suffix = fmt.Sprintf("%s?%s", parseSuffix.Path, query.Encode())
 	}
 
-	return fmt.Sprintf("%s%s", c.config.BaseURL, suffix)
+	if c.config.APIType == APITypeAzure || c.config.APIType == APITypeAzureAD {
+		azureDeploymentName := c.config.GetAzureDeploymentByModel(args.model)
+		if azureDeploymentName == "" {
+			azureDeploymentName = "UNKNOWN"
+		}
+		baseURL = fmt.Sprintf("%s/%s", baseURL, azureAPIPrefix)
+		if containsSubstr(azureDeploymentsEndpoints, suffix) {
+			baseURL = fmt.Sprintf("%s/%s/%s", baseURL, azureDeploymentsPrefix, azureDeploymentName)
+		}
+	}
```

Comment:
This part is pretty hard to figure about 🤔  Could I suggest breaking it down into more functions and keeping `fullURLOptions` closer to the code that uses it?

Something like this:

```go
func (c *Client) suffixWithAPIVersion(previousSuffix string) (newSuffix string) {
	parsedSuffix, err := url.Parse(previousSuffix)
	if err != nil {
		panic("failed to parse url suffix")
	}

	query := parsedSuffix.Query()
	query.Add("api-version", c.config.APIVersion)
	return fmt.Sprintf("%s?%s", parsedSuffix.Path, query.Encode())
}

func (c *Client) baseURLWithAzureDeployment(baseURL, suffix, model string) (newBaseURL string) {
	azureDeploymentName := c.config.GetAzureDeploymentByModel(model)
	if azureDeploymentName == "" {
		azureDeploymentName = "UNKNOWN"
	}
	baseURL = fmt.Sprintf("%s/%s", baseURL, azureAPIPrefix)
	if containsSubstr(azureDeploymentsEndpoints, suffix) {
		baseURL = fmt.Sprintf("%s/%s/%s", baseURL, azureDeploymentsPrefix, azureDeploymentName)
	}
	return baseURL
}


// fullURL returns full URL for request.
func (c *Client) fullURL(suffix string, setters ...fullURLOption) string {
	baseURL := strings.TrimRight(c.config.BaseURL, "/")
	urlOptions := fullURLOptions{}
	for _, setter := range setters {
		setter(&args)
	}

	if c.config.APIType == APITypeAzure || c.config.APIType == APITypeAzureAD {
		baseURL = baseURLWithAzureDeployment(baseURL, suffix, urlOptions.model)
	}

	if c.config.APIVersion != "" {
		suffix = c.suffixWithAPIVersion(suffix)
	}

	return fmt.Sprintf("%s%s", baseURL, suffix)
}
```

Reply:
I've completed the suggested changes. Could you please review it again?